### ISSUE TITLE
[Bug]: LC0031 - LockTable still used in table triggers

### ISSUE THREAD

**Issue by NKarolak** [ALCops/Analyzers#530]
### Analyzer

LinterCop

### Issue type

False positive (diagnostic fires when it shouldn't)

### Rule ID

LC0031

### Description

I am not sure whether it is a false positive, but maybe you know.

LC0031 fires too when LockTable is used on `Rec` in table triggers. These LockTable patterns are still found in standard code, but I never saw any instruction on how to replace those with ReadIsolation (or anything else). The ALCops code action does not update anything here.

This is where you find LockTable in standard code until today, just examples:

`OnDelete`

IssuedFinChargeMemoHeader.Table.al
```
    trigger OnDelete()
    begin
        TestField("No. Printed");
        LockTable();
        FinChrgMemoIssue.DeleteIssuedFinChrgLines(Rec);

        FinChrgCommentLine.SetRange(Type, FinChrgCommentLine.Type::"Issued Finance Charge Memo");
        FinChrgCommentLine.SetRange("No.", "No.");
        FinChrgCommentLine.DeleteAll();
    end;
```

PurchCrMemoHdr.Table.al
```
    trigger OnDelete()
    var
        PostedDeferralHeader: Record "Posted Deferral Header";
        PostPurchDelete: Codeunit "PostPurch-Delete";
    begin
        PostPurchDelete.IsDocumentDeletionAllowed("Posting Date");
        LockTable();
        PostPurchDelete.DeletePurchCrMemoLines(Rec);

        PurchCommentLine.SetRange("Document Type", PurchCommentLine."Document Type"::"Posted Credit Memo");
        PurchCommentLine.SetRange("No.", "No.");
        PurchCommentLine.DeleteAll();

        ApprovalsMgmt.DeletePostedApprovalEntries(RecordId);
        PostedDeferralHeader.DeleteForDoc(
            "Deferral Document Type"::Purchase.AsInteger(), '', '',
            PurchCommentLine."Document Type"::"Posted Credit Memo".AsInteger(), "No.");
    end;
```

And you find it a lot in `OnInsert` triggers, such as in CashFlowWorksheetLine.Table.al:
```
    trigger OnInsert()
    begin
        LockTable();
        Rec.ValidateShortcutDimCode(1, "Shortcut Dimension 1 Code");
        Rec.ValidateShortcutDimCode(2, "Shortcut Dimension 2 Code");
    end;
```

ResJournalBatch.Table.al
```
    trigger OnInsert()
    begin
        LockTable();
        ResJnlTemplate.Get("Journal Template Name");
    end;
```

ResJournalLine.Table.al
```
    trigger OnInsert()
    begin
        LockTable();
        ResJnlTemplate.Get("Journal Template Name");
        ResJnlBatch.Get("Journal Template Name", "Journal Batch Name");

        Rec.ValidateShortcutDimCode(1, "Shortcut Dimension 1 Code");
        Rec.ValidateShortcutDimCode(2, "Shortcut Dimension 2 Code");
    end;
```

**Comment by Arthurvdv** [comment]
> **AI investigation report.** This comment was researched and written by Claude Code (AL SDK source, Microsoft Learn, microsoft/BCQuality, community blogs and the W1 app sources from BC 23.5 to 28.4), reviewed by the maintainer before posting.

Thanks for the detailed report. We dug into this from four directions before deciding, and the conclusion is: LC0031 firing on `LockTable()` inside table triggers is intended, but the code action not doing anything on the bare `LockTable()` form is a real bug, and we will fix it.

**Why the trigger case is reported**

- The compiler binds bare `LockTable()`, `Rec.LockTable()`, `this.LockTable()` and `MyVar.LockTable()` to one and the same built-in method (`TableClassTypeSymbol` in the AL SDK); there is no self-receiver special case, and LockTable is not deprecated there either.
- The runtime effect is the same everywhere: LockTable takes no lock itself, it marks the table so that every subsequent read of that table in the transaction, on any variable, uses UPDLOCK, and it disables tri-state optimistic reads for that table (Microsoft Learn: [performance-developer](https://learn.microsoft.com/en-us/dynamics365/business-central/dev-itpro/performance/performance-developer) and [devenv-tri-state-locking](https://learn.microsoft.com/en-us/dynamics365/business-central/dev-itpro/developer/devenv-tri-state-locking)). `ReadIsolation` is per record instance. That difference is the reason the rule exists, and it is exactly the same inside `OnInsert`.
- Microsoft's own guidance (microsoft/BCQuality [do-not-locktable-in-read-only-procedure.md](https://github.com/microsoft/BCQuality/blob/main/microsoft/knowledge/performance/do-not-locktable-in-read-only-procedure.md) and [prefer-readisolation-over-locktable-for-reads.md](https://github.com/microsoft/BCQuality/blob/main/microsoft/knowledge/performance/prefer-readisolation-over-locktable-for-reads.md)) endorses LockTable only for a read that a directly following Insert/Modify/Delete depends on.

**What the Base App actually does with it**

We scanned the W1 app sources from BC 23.5 to 28.4. `LockTable()` in table `OnInsert`/`OnModify`/`OnDelete`/`OnRename` bodies went from 80 to 78: two were removed (e.g. `Item Journal Line`.OnInsert, deleted outright, no ReadIsolation added), none were added. 77 of the 78 are in Base Application; Business Foundation, E-Document Core, Subscription Billing and Excise Taxes contain no LockTable at all. The triggers you quoted are the two big groups: `LockTable(); ResJnlTemplate.Get(...)` never reads the Res. Journal Line table again, so the call only re-enables pessimistic reads for the rest of the transaction; the posted-document `OnDelete` locks the header while deleting lines. Where Microsoft does need "make sure nobody else took this number" in `OnInsert`, new code uses `ReadIsolation` on a second instance instead (`Customer.Table.al`: `Customer.ReadIsolation(IsolationLevel::ReadUncommitted); while Customer.Get("No.") do ...`; same shape in the 2025 Sustainability ESG tables).

**What to do with a trigger `LockTable()`**

- If no read of the same table follows that a write depends on: delete the line.
- If one does (for example `Xxx2.FindLast()` for the next entry number): put `Xxx2.ReadIsolation(IsolationLevel::UpdLock)` on the instance that reads.

**What changes in ALCops**

- The code action now also rewrites bare `LockTable();` (to `ReadIsolation(IsolationLevel::UpdLock);`). It keeps converting rather than deleting, because whether the call is dead is a judgement the developer has to make; the docs page explains the choice.
- The LC0031 page on alcops.dev gets a "LockTable in table triggers" section with the above.

---
🤖 Created by [Claude Code](https://claude.com/claude-code)

**Comment by Arthurvdv** [comment]
@NKarolak I always thought that a `LockTable()` could be replaced with `ReadIsolation(IsolationLevel::UpdLock)` without any consequences, which then would allow to benefit from the tri-state locking feature. I’ve also asked Claude (see above) to challenge this and so far it hasn’t found a reason why we shouldn’t make that change.

In the BaseApp, I indeed see still quite a lot of old-school `LockTable()` usages. Ideally, this would be a simple search-and-replace, but I know Microsoft tends to be reserved about making changes in the BaseApp on existing code. If I'm not mistaken there are still some custom BaseApps out there where some changes can make it difficult for partners to merge new changes into their customized versions. Because of that I think it would be better to look at newer apps/extensions outside the BaseApp, which would give a more representative picture of how `ReadIsolation()` is being used.

I also noticed that the CodeFix on these examples seems to be broken. I’ll create a PR to fix that.

**Comment by Arthurvdv** [comment]
Fixed in #547 (code fix handles the bare `LockTable()` form; the diagnostic in table triggers stays by design, see the comment above). Docs: https://alcops.dev/docs/analyzers/lintercop/lc0031/ (ALCops/alcops.dev#185, #186).
